### PR TITLE
add REST API command line interface 'rest-cli'.

### PR DESCRIPTION
**Pull request by MusclePr**
## Context 

Like rcon-cli, it supports a REST API command line interface.

## Choices

Since rcon-cli is convenient, I also tried imitating rest-cli.

## Test instructions

1. download test script.
[test-rest-cli.zip](https://github.com/thijsvanloef/palworld-server-docker/files/15074737/test-rest-cli.zip)

2. extract to project root.
```shell
unzip test-rest-cli.zip
chmod +x test-rest-cli.sh
```
4. build docker image with debug tag.
```shell
docker build -t thijsvanloef/palworld-server-docker:debug .
```
5. run test-rest-cli.sh and connect server from client manualy.
```
$ ./test-rest-cli.sh 
[+] Running 1/2
 ⠸ Network palworld-server-docker_default  Created
  ✔ Container palworld-server               Started 
wait healthy...... done.
TEST ... docker exec -i palworld-server rest-cli
OK
TEST ... docker exec -i palworld-server rest-cli -h
OK
TEST ... docker exec -i palworld-server rest-cli announce -h
OK
TEST ... docker exec -i palworld-server rest-cli ban -h
OK
TEST ... docker exec -i palworld-server rest-cli unban -h
OK
TEST ... docker exec -i palworld-server rest-cli kick -h
OK
TEST ... docker exec -i palworld-server rest-cli shutdown -h
OK
TEST ... docker exec -i palworld-server rest-cli unknown foo bar
OK
TEST ... docker exec -i palworld-server rest-cli info
OK
TEST ... docker exec -i palworld-server rest-cli settings
OK
TEST ... docker exec -i palworld-server rest-cli metrics
OK
SERVER ADDRESS: 172.21.57.49:8211
TEST ... docker exec -i palworld-server rest-cli players
wait login..OK
TEST ... docker exec -i palworld-server rest-cli announce "Hello PalWorld!"
OK
TEST ... docker exec -i palworld-server rest-cli announce '{"message":"Logout to continue tests."}'
OK
TEST ... cat json-file | docker exec -i palworld-server rest-cli announce -
OK
TEST ... docker exec -i palworld-server rest-cli save
OK
TEST ... docker exec -i palworld-server rest-cli shutdown 300
OK
wait shutdown or logout......logout to stop now
TEST ... docker exec -i palworld-server rest-cli stop
OK
... done.
TEST ... docker exec -i palworld-server rest-cli info
OK
[+] Running 2/2
 ✔ Container palworld-server               Removed
 ✔ Network palworld-server-docker_default  Removed
 ```

## Checklist before requesting a review

- [x] I have performed a self-review/test of my code
- [x] I've added documentation about this change to the [docs](https://github.com/thijsvanloef/palworld-server-docker/tree/main/docusaurus/docs).
- [x] I've not introduced breaking changes.
- [x] My changes do not violate linting rules.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -155,7 +155,8 @@ COPY ./scripts /home/steam/server/
 RUN chmod +x /home/steam/server/*.sh && \
     mv /home/steam/server/backup.sh /usr/local/bin/backup && \
     mv /home/steam/server/update.sh /usr/local/bin/update && \
-    mv /home/steam/server/restore.sh /usr/local/bin/restore
+    mv /home/steam/server/restore.sh /usr/local/bin/restore && \
+    ln -sf /home/steam/server/rest_api.sh /usr/local/bin/rest-cli
```
Comment (MusclePr) [inline Dockerfile:159]: Symbolic links are just a preference.

File: scripts/helper_functions.sh [hunk 2]
```diff
@@ -176,7 +176,8 @@ REST_API(){
   local URL="http://localhost:${REST_API_PORT}/v1/api/${1}"
   local ACCEPT="Accept: application/json"
   local USERPASS="admin:${ADMIN_PASSWORD}"
-  if [ "${DATA}" = "" ]; then
+  local post_api="save|stop"
+  if [ "${DATA}" = "" ] && [[ ! ${1} =~ ${post_api} ]]; then
```
Comment (MusclePr) [inline scripts/helper_functions.sh:180]: "save" and "stop" require the method to be POST.

File: scripts/rest_api.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      json="{\"waittime\":${sec},\"message\":\"${msg}\"}"
+      ;;
+    "unban")
+      if [[ ${json} =~ ${help} ]]; then
+        echo "Usage: $(basename ${0}) ${api} <steam_00000000000000000>"
+        exit 1
+      fi
+      json="{\"userid\":\"${2}\"}"
+      ;;
+    esac
```
Comment (MusclePr) [inline scripts/rest_api.sh:84]: If it is not JSON, it will be parsed as a command line parameter.
It simply converts the parameters to JSON for each API.

**Comment by MusclePr** [comment]
Due to an increase in minor fixes, I will temporarily organize the commits.

File: README.md [hunk 4]
```diff
@@ -333,6 +333,83 @@ This will open a CLI that uses RCON to write commands to the Palworld Server.
...
+| metrics                          | Get the server metrics.                             |
+| announce                         | Announce message.                                   |
+| kick {SteamID}                   | Kick player.                                        |
+| ban {SteamID}                    | Ban player.                                         |
+| unban {SteamID}                  | Unban player.                                       |
+| save                             | Save the world.                                     |
+| shutdown {Seconds} {MessageText} | Shutdown the server                                 |
+| stop                             | Force stop the server.                              |
+
+For a official documents go to: [https://tech.palworldgame.com/category/rest-api](https://tech.palworldgame.com/category/rest-api)
```
Comment (MusclePr) [inline README.md:411]: a official -> an official

File: scripts/rest_api.sh [hunk 5]
```diff
@@ -0,0 +1,92 @@
+#!/bin/bash
+SCRIPT=$(basename "${0}")
+ORGPATH=$(readlink -fn "${0}")
+ORGDIR=$(dirname "${ORGPATH}")
+#shellcheck source=scripts/helper_functions.sh
+source "${ORGDIR}/helper_functions.sh"
```
Comment (MusclePr) [inline scripts/rest_api.sh:6]: The symbolic link /usr/local/bin/rest-cli refers to helper_functions.sh in the same location as the original rest_api.sh.

File: scripts/rest_api.sh [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      json="{\"waittime\":${sec},\"message\":\"${msg}\"}"
+      ;;
+    "unban")
+      if [[ ${json} =~ ${help} ]]; then
+        echo "${usage} <steam_00000000000000000>"
+        exit 1
+      fi
+      json="{\"userid\":\"${2}\"}"
+      ;;
+    esac
```
Comment (MusclePr) [inline scripts/rest_api.sh:88]: Here we are just converting the CLI parameters to JSON when the parameters are not JSON.

File: scripts/rest_api.sh [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  fi
+fi
+
+REST_API "${api}" "${json}" && echo ""
```
Comment (MusclePr) [inline scripts/rest_api.sh:92]: I added a line break because the output JSON ends with "}".

**Comment by thijsvanloef** [comment]
Hi @MusclePr ! Thanks again for the REST API PR's! Will take a look at it tonight

**Comment by MusclePr** [comment]
Fixed a part of README.md.

**Comment by thijsvanloef** [comment]
Sorry I took so long, got sidetracked, merging!

**Pull request merged**